### ISSUE TITLE
_ready() is called twice when instancing a tool in the editor

### ISSUE THREAD

**Issue by Zylann** [godotengine/godot#6005]
Tested on Godot 2.1 rc1 windows 10 64 bits

I make an editor plugin in which I instantiate a scene that contains a control for a container in the editor.
However, initialization fails compeletely because _ready() is called twice, and I cannot rely on _enter_tree().

Repro:
[PluginReadyTwice.zip](https://github.com/godotengine/godot/files/395081/PluginReadyTwice.zip)
This project contains a dummy plugin that instances a custom control. In `_ready()`, this control prints a message.
Open the project, see the message print twice in the console.

Note: `_ready()` is called once normally for games.

Note 2: I also don't understand why the doc recommends using `_enter_tree()` for initialization, because it makes `onready()` and `get_node()` useless.

Note 3: unrelated, but I get countless errors in the console when developping my tool since rc1 is out. They don't seem harmful, but I didn't managed to get them 100% reproductible in a simple project yet :s

**Comment by akien-mga** [comment]
I have a similar issue which is likely a duplicate: `_init` and `_ready` are also called twice for AutoLoads (regardless of their Singleton status).

**Comment by gau-veldt** [comment]
I just hit the ready_ twice issue myself.  Here's a minimal [ready_x2.tar.gz](https://github.com/godotengine/godot/files/426136/ready_x2.tar.gz) project to the effect.  It clearly shows the single call (when brought in from main) versus the double call (when brought in from an autoload).  It also shows the main script and autoload themselves only get _ready once but the object instantiated is getting a differing number of _ready calls.  In my case I'm attaching from an external script but I still think it's a duplicate of the above issue.

**PS:** As a workaround I define a member var as a flag to enforce a one-off initialization:
```
var init=false
func _ready():
    if !init:
        init=true
        # only runs once
```

**Comment by vnen** [comment]
This thing is quite annoying. I dug a little but have no idea of what is causing it. I noticed that `_ready()` is called twice on plugins but only when the editor starts, not when disabling/re-enabling it. So I guess there's something on in the initialization that propagates the `NOTIFICATION_READY` one extra time.

**Comment by reduz** [comment]
Its probably related to your window being moved arond the docks? I don't
think it's a bug in itself

**Comment by reduz** [comment]
Maybe for Godot 3.0 we could make _ready only run once, and add something
else for when reentering

**Comment by reduz** [comment]
Or just leave enter tree for that..

**Comment by vnen** [comment]
> Its probably related to your window being moved arond the docks? I don't
think it's a bug in itself

No, it runs `_ready()` twice just by opening the editor. It's not even a dock plugin, it happens on import and main screen plugins too.

In a game you have control. If you never reparent anything, you can be sure `_ready()` is only called once. In the editor you have to be a bit more careful, but you should have control of the scenes you instance and add to the tree manually.

However, the problem here is that `_ready()` is a actually called twice, it's not being reparented. In my plugin I create a new instance every time and add it to the tree once, so I shouldn't have two calls for `_ready()`. Based on the prints, is the same instance. Again: this only happens when starting the editor, not when re-enabling the plugin.

The problem with singletons is similar (likely the same cause). The effect is worse though: you have to *always* make initialization guards for a **singleton**, which by definition should be added to tree only once.

**Comment by volzhs** [comment]
I think I found problem.
I happens even not editor plugin.

**main.tscn**
```gdscript
extends Control
func _enter_tree():
	var sp = load("res://Sprite.tscn").instance()
	add_child(sp)
```
**Sprite.tscn**
```gdscript
extends Sprite
func _ready():
	print("sprite ready")
```

I can see `sprite ready` twice on output when run.
to work around this, use `call_deferred("add_child",sp)` instead.

EditorPlugin example also add a scene in `_enter_tree`
So, I made `call_deffered` instead of just call adding it.

```gdscript
func _enter_tree():
    dock = preload("res://addons/my_custom_dock/my_dock.tscn").instance()
    call_deffered("add_control_to_dock", DOCK_SLOT_LEFT_UL, dock)
```

AND PROBLEM IS GONE!

**Comment by volzhs** [comment]
I think, the REAL problem is `_ready` is called twice when added in `_enter_tree`.

**Comment by bojidar-bg** [comment]
@volzhs Nice bug summary, though I hope it isn't a too paranormal issue... :+1: 

So, I made a quick check for when _ready is emitted -- seems like it is only here:
[scene/main/node.cpp:175](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L175)
```cpp
void Node::_propagate_ready() {

	data.blocked++;
	for (int i=0;i<data.children.size();i++) {

		data.children[i]->_propagate_ready();
	}
	data.blocked--;
	notification(NOTIFICATION_READY); // <---- here

}
```

Then... debugging everything on a sheet of "paper" produces this:

Node | "event"
---- | ------
`/root/editor` | ... → `add_child(plugin)`
`/root/editor/plugin` | [`_set_tree`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L1375) → [`_propagate_enter_tree`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L2664)
`/root/editor/plugin` | script (`_enter_tree`) → `add_child(my_node)` 
`/root/editor/plugin/my_node` | [`_set_tree`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L1375) → [`_propagate_enter_tree`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L2664) & [`_propagate_ready`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L2665)
`/root/editor/plugin/my_node` | `_propagate_ready` → script(`_ready`) :exclamation:
`/root/editor/plugin` | `_propagate_enter_tree` (Continuation) → [children*`_propagate_enter_tree`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L219)
`/root/editor/plugin` | `_set_tree` (Continuation)→ [`_propagate_ready`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L2665) → [children*`_propagate_ready`](https://github.com/godotengine/godot/blob/master/scene/main/node.cpp#L219)
`/root/editor/plugin/my_node` | parent → `_propagate_ready` → script(`_ready`) :exclamation:

Or, in short, when the child node is added, it is immediately `_ready`-ed, then all the children of the node that added it are readied...

To fix this, one of the two "paths" which call `_ready` should be stopped somehow. I guess the simplest way to do this (and remain mostly backwards-compatible) would be to not call `children._ready` on children added in `parent._enter_tree`, but wait for `parent._ready` to propagate it.
To do this, we might add `data.ready = false`, which we set to `true` when we start propagating `_ready` and reset it when we exit the tree (until 3.0). Then, we would check/pass the parent's flag in `_set_tree`, and not call `_propagate_ready` in this case.

Hopefully, I was clear enough... :smiley:

**Comment by boruok** [comment]
Issue **persist** v3.4.1.rc1.official [7b0801c7f]

**Comment by Calinou** [comment]
> Issue **persist** v3.4.1.rc1.official [[7b0801c](https://github.com/godotengine/godot/commit/7b0801c7fb4416625fb9ca124b41b93677689420)]

This may be related to https://github.com/godotengine/godot/issues/47731 or https://github.com/godotengine/godot/issues/31002, which means it would be fixed in `master` already.
